File: examples/typescript-express-zero-config/src/user-service.ts
export interface User {
  id: string;
  name: string;
  email: string;
  createdAt: Date;
}

export class UserService {
  private users: Map<string, User> = new Map();
  private nextId = 1;

  public findAll(): User[] {
    return Array.from(this.users.values());
  }

  public findById(id: string): User | undefined {
    return this.users.get(id);
  }

  public create(data: Omit<User, "id" | "createdAt">): User {
    const id = String(this.nextId++);
    const user: User = {
      id,
      name: data.name,
      email: data.email,
      createdAt: new Date(),
    };
    this.users.set(id, user);
    return user;
  }

  public update(id: string, data: Partial<Omit<User, "id" | "createdAt">>): User | undefined {
    const existing = this.users.get(id);
    if (!existing) return undefined;
    const updated: User = { ...existing, ...data };
    this.users.set(id, updated);
    return updated;
  }

  public delete(id: string): boolean {
    return this.users.delete(id);
  }
}

File: examples/go-rest-zero-config/services/user_service.go
package services

import (
	"fmt"
	"sync"
	"time"
)

// User represents a user in the system.
type User struct {
	ID        string `json:"id"`
	Name      string `json:"name"`
	Email     string `json:"email"`
	CreatedAt string `json:"created_at"`
}

// UserService manages user CRUD operations with in-memory storage.
type UserService struct {
	mu     sync.RWMutex
	users  map[string]User
	nextID int
}

// NewUserService creates a new UserService with an empty store.
func NewUserService() *UserService {
	return &UserService{
		users:  make(map[string]User),
		nextID: 1,
	}
}

// FindAll returns all users.
func (s *UserService) FindAll() []User {
	s.mu.RLock()
	defer s.mu.RUnlock()

	result := make([]User, 0, len(s.users))
	for _, u := range s.users {
		result = append(result, u)
	}
	return result
}

// FindByID looks up a user by their unique identifier.
func (s *UserService) FindByID(id string) (User, bool) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	u, ok := s.users[id]
	return u, ok
}

// Create adds a new user and returns the created record.
func (s *UserService) Create(name, email string) User {
	s.mu.Lock()
	defer s.mu.Unlock()

	id := fmt.Sprintf("%d", s.nextID)
	s.nextID++

	u := User{
		ID:        id,
		Name:      name,
		Email:     email,
		CreatedAt: time.Now().UTC().Format(time.RFC3339),
	}
	s.users[id] = u
	return u
}

// Update modifies an existing user's fields. Returns false if the user does not exist.
func (s *UserService) Update(id, name, email string) (User, bool) {
	s.mu.Lock()
	defer s.mu.Unlock()

	u, ok := s.users[id]
	if !ok {
		return User{}, false
	}
	u.Name = name
	u.Email = email
	s.users[id] = u
	return u, true
}

// Delete removes a user by ID. Returns true if the user existed.
func (s *UserService) Delete(id string) bool {
	s.mu.Lock()
	defer s.mu.Unlock()

	_, ok := s.users[id]
	if ok {
		delete(s.users, id)
	}
	return ok
}
